doc/languages-frameworks/maven.md

```diff
@@ -0,0 +1,298 @@
+
+> This should be executed in the project's source repository or be told which _pom.xml_ to execute with.
+
+```bash
+# run this step within the project's source repository
+❯ mvn org.nixos.mvn2nix:mvn2nix-maven-plugin:mvn2nix
+
+❯ cat project-info.json | jq | head
+{
+  "project": {
+    "artifactId": "maven-demo",
+    "groupId": "org.nixos",
+    "version": "1.0",
+    "classifier": "",
+    "extension": "jar",
+    "dependencies": [
+      {
+        "artifactId": "maven-resources-plugin",
+```
+
+This file is then given to the _buildMaven_ function, and it returns 3 properties.
+
+**repo**:
+    A Maven repository that is a symlink farm of all the dependencies found in the _project-info.json_
+
+
+**build**:
+    A simple derivation that runs through _mvn compile_ & _mvn package_ to build the JAR. You may use this as inspiration for more complicated derivations.
+
+Here is an example of building the Maven repository
+```nix
+{ pkgs ? import <nixpkgs> {} }:
+with pkgs;
+let build = buildMaven ./project-info.json;
+in
+build.repo
+```
+
+The benefit over the _double invocation_, is that the _/nix/store_ entry is a _linkFarm_ of every package, so that changes to your dependency set doesn't involve downloading everything from scratch.
+
+```bash
+❯ tree /nix/store/g87va52nkc8jzbmi1aqdcf2f109r4dvn-maven-repository | head
+/nix/store/g87va52nkc8jzbmi1aqdcf2f109r4dvn-maven-repository
+├── antlr
+│   └── antlr
+│       └── 2.7.2
+│           ├── antlr-2.7.2.jar -> /nix/store/d027c8f2cnmj5yrynpbq2s6wmc9cb559-antlr-2.7.2.jar
+│           └── antlr-2.7.2.pom -> /nix/store/mv42fc5gizl8h5g5vpywz1nfiynmzgp2-antlr-2.7.2.pom
+├── avalon-framework
+│   └── avalon-framework
+│       └── 4.1.3
+│           ├── avalon-framework-4.1.3.jar -> /nix/store/iv5fp3955w3nq28ff9xfz86wvxbiw6n9-avalon-framework-4.1.3.jar
+```
+
+## Building a JAR
+
+Regardless of which strategy is chosen above, the step to build the derivation now becomes the same.
+
+```nix
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, lib ? pkgs.lib
+, maven ? pkgs.maven }:
+# pick a repository derivation from above
+let repository = ...;
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = lib.cleanSourceWith {
+    filter = lib.cleanSourceFilter;
+    src = lib.cleanSourceWith {
+      filter = path: type:
+        !(lib.pathIsDirectory (path) && baseNameOf (toString (path))
+          == "target");
+      src = ./.; # or fetchFromGitHub, cleanSourceWith, etc
+    };
+  };
+  buildInputs = [ maven ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    mkdir $out/share/java
+    cp target/${pname}-${version}.jar $out/share/java
+  '';
+}
+```
+
+> We place the library in _$out/share/java_ since JDK have a _stdenv setup hook_ that add any JARs in the _share/java_ directories of the build inputs to the CLASSPATH environment.
```

Comment:
```suggestion
> We place the library in _$out/share/java_ since JDK package has a _stdenv setup hook_ that adds any JARs in the _share/java_ directories of the build inputs to the CLASSPATH environment.
```